V3 add classes (#2035)
Vyxal/Vyxal: documentation/Advanced Features/Objects and Extensions.md

# Objects and Extension Methods

Vyxal is already turing complete, and can be used to easily solve many problems.
However, say there was a situation where you needed to model a complex real-world
object, such as a car. You could use two lists to map keys to values, but that
would require making sure that the two lists are always in sync. Objects take
away this hassle, and allow you to model real-world objects in a way that is
convienient and easy to use.

## The OOP Model

Unlike most OOP languages, Vyxal does not have traditional classes. Instead, it
has struct-like objects. These objects contain only data, and no methods. To add
methods to an object, you can use extension methods (covered later).

There is also no inheritance, meaning that composition is the only way to reuse
code. This is not a bad thing, as it makes the implementation of objects much
simpler.

## Defining Objects

> To define an object:

Reviewer: The terminology has been confusing me a bit. What you define with `#:O` is a struct, right? Not a prototype like JS? Because calling it an object implies it's a JS-like object that acts as a prototype for other objects, rather than a C-struct-like template for making objects. I feel like it'd be better to refer to say `#:O StructName` rather than `#:O ObjectName`, because `struct` is something a lot of us are familiar with, and is basically the same concept, as far as I can tell

Author: You're right that it's more like a struct, but I wanted to avoid calling it that so as to not create confusion with the terminology used for control flow structures. Maybe there's a better term available that I don't know of, like record.

Reviewer: Record sounds good to me. I think record might imply all members are public to some, but whatever

```

> #:O ObjectName | #$restrictedMember #=privateMember <data> #$restrictedMemberWithValue
> <data> #=privateMemberWithValue #!publicMember}

Reviewer: ```suggestion
#:O ObjectName |
  #$restrictedMember
  #=privateMember
  <data> #$restrictedMemberWithValue
  <data> #=privateMemberWithValue
  #!publicMember}
```
Not a big deal (you don't have to accept this) but it might be clearer (I was confused at first because I thought the first `<data>` belonged to the first `#=privateMember` before realizing this is a stack-based language)

members are typically a bad idea. It doesn't have the same meaning as `#!` for
constants, but the `!` was appropriate to signify the danger of public members.

Members without a default value will be set from the stack when the object is
created.

## Using Objects

To create a new object:

```
#$ObjectName Ė
```

Defining an object creates a variable with the same name as the object. This
pushes a constructor for the object onto the stack. The `Ė` element calls the
constructor and pushes the created object onto the stack. This is analogous to
calling a function, which is intentional - there's no need for a `new` keyword
when you can just call the constructor in a similar way to a function.

To access a member of an object:

```
## Assuming the object is on the stack
"memberName" «
```

This pushes the value of the member onto the stack. If the member is a private
member, this will fail, unless inside an extension method.

To set a member of an object:

```
## Assuming the object is on the stack  
"memberName" <value> ŀ
```

This sets the value of the member to the value on the stack. If the member is a

> restriced or private member, this will fail, unless inside an extension method.

Reviewer: ```suggestion
restricted or private member, this will fail, unless inside an extension method.

```

Maybe also say that extension methods will be covered later, and link to them

* `type` is the type of an argument to the extension method. This can be any valid type name, as well as `*`, which indicates an any type.
* `implementation` is the implementation of the extension method. This will always
be the last branch.

To use an extension method, call it as you would any custom element:

```
#:@symbol
```

### Important Notes

* Extension methods require at least one argument. Attempting to define an
extension method with no arguments will result in an error. Use the define
structure instead.
* Extension methods will be scoped globally. Currently, there is no way to define
an extension method that is only available within a certain scope/context.
* The types of the arguments are checked at runtime. If the types do not match,
an error will be thrown.
* Like custom elements, the implementation branch is where the actual code for the element or modifier is written. When called, the method will pop all arguments from the stack
and execute the implementation as if it were a lambda. The implementation branch can be any valid Vyxal code.

### Order of Extension Method Matching

When a symbol is called, first the interpreter will check for an extension method

> with the appropriate types of items on the stack. If found, it will execute the

Reviewer: Might be worth mentioning the phrases "multimethods" or "multiple dispatch" somewhere, to help people who already understand those concepts (as far as I can tell, "multiple dispatch" is an accurate way of describing the type-based overloading Vyxal does). Maybe also throw in a bit about *why* we do type-based overloading (better to have the language do it for you than do it manually, safer than languages like Python which let you do whatever, blah blah blah). Maybe mention Julia?

extension. Otherwise, the interpreter will check for a custom element with the
same symbol. If found, it will execute the custom element. Otherwise, it will
finally try a lookup of the built-in symbols. If none of these are found, an
error will be thrown.

```
Highest Priority
- Extension Methods
- Custom Elements
- Built-in Symbols
Lowest Priority
```

## Example

```
#:O Map | 
  #[#] #!keys
  #[#] #!values
}

#:>> put | value | * | key | * | mp | Map | 
  #$mp "keys" ᵇ« #$key & ŀ 
  #$mp "values" ᵇ« #$value & ŀ
  #$mp
}

#$Map Ė
"key1" "value1" #:@put
```

This example is the same as the previous example, except that it uses an extension
method instead of a lambda. Calling `put` with any other value for `mp` that is

> not a `Map` object will not execute the extension method.

Reviewer: ```suggestion
not a `Map` object will throw an error since no `put` method has been
defined for that value's type.

```

## Literate Mode

To define an object:

```
object ObjectName =>
  $restrictedMember
  :=privateMember
  :!=publicMember
}
```

To create an object:

```
`ObjectName`
## or
$ObjectName call
```

To access a member of an object:

```
"memberName" @<=
```

To set a member of an object:

```
"memberName" value @=>
```

To define an extension method:

```
extension symbol given
  argA as typeA,
  argB as typeB,
  ...
  argN as typeN
does
    implementation
}
```

> _Note that any branch keyword can be used in place of `does`, `as`, `given` and `,`._

Reviewer: ```suggestion
> [!note]
> Any branch keyword can be used in place of `does`, `as`, `given` and `,`._

```
This isn't necessary but it's kinda cool. Here's what it looks like:

> [!note]
> Any branch keyword can be used in place of `does`, `as`, `given` and `,`._
